File: coding-interview-patterns/python/patterns/prefix_sums.py
from collections import defaultdict
def subarray_sum_k(nums,k):
    pref=0; seen=defaultdict(int); seen[0]=1; ans=0
    for x in nums:
        pref+=x; ans+=seen[pref-k]; seen[pref]+=1
    return ans

class NumMatrix:
    def __init__(self, M):
        m=len(M); n=len(M[0]) if m else 0
        self.P=[[0]*(n+1) for _ in range(m+1)]
        for i in range(1,m+1):
            for j in range(1,n+1):
                self.P[i][j]=M[i-1][j-1]+self.P[i-1][j]+self.P[i][j-1]-self.P[i-1][j-1]
    def sumRegion(self,r1,c1,r2,c2):
        P=self.P; return P[r2+1][c2+1]-P[r1][c2+1]-P[r2+1][c1]+P[r1][c1]

def apply_range_adds(n, updates):
    diff=[0]*(n+1)
    for l,r,v in updates:
        diff[l]+=v
        if r+1<len(diff): diff[r+1]-=v
    cur=0; out=[0]*n
    for i in range(n):
        cur+=diff[i]; out[i]=cur
    return out

File: coding-interview-patterns/scala/src/main/scala/patterns/PrefixSums.scala
package patterns
import scala.collection.mutable
object PrefixSums {
  def subarraySumK(nums:Array[Int], k:Int): Int = {
    val map = mutable.Map[Int,Int](0->1)
    var pref=0; var ans=0
    for (x <- nums){ pref+=x; ans += map.getOrElse(pref-k,0); map.update(pref, map.getOrElse(pref,0)+1) }
    ans
  }
  class NumMatrix(M:Array[Array[Int]]){
    private val m=M.length; private val n=if (m==0) 0 else M(0).length
    private val P = Array.ofDim[Int](m+1, n+1)
    for (i<-1 to m; j<-1 to n) P(i)(j)=M(i-1)(j-1)+P(i-1)(j)+P(i)(j-1)-P(i-1)(j-1)
    def sumRegion(r1:Int,c1:Int,r2:Int,c2:Int): Int =
      P(r2+1)(c2+1)-P(r1)(c2+1)-P(r2+1)(c1)+P(r1)(c1)
  }
  def applyRangeAdds(n:Int, updates:Array[Array[Int]]): Array[Int] = {
    val diff = Array.fill(n+1)(0)
    for (u <- updates){ val l=u(0); val r=u(1); val v=u(2); diff(l)+=v; if (r+1<diff.length) diff(r+1)-=v }
    val out = Array.fill(n)(0); var cur=0
    for (i<-0 until n){ cur+=diff(i); out(i)=cur }
    out
  }
}
